File: frontend/src/app/services/commande-achat.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { CommandeAchat } from '../models/commande-achat.model';
import { environment } from '../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class CommandeAchatService {
  private apiUrl = `${environment.apiUrl}/commandes`;

  constructor(private http: HttpClient) { }

  getAllCommandes(): Observable<CommandeAchat[]> {
    return this.http.get<CommandeAchat[]>(this.apiUrl);
  }

  getCommandeById(id: number): Observable<CommandeAchat> {
    return this.http.get<CommandeAchat>(`${this.apiUrl}/${id}`);
  }

  createCommande(commande: CommandeAchat): Observable<CommandeAchat> {
    return this.http.post<CommandeAchat>(this.apiUrl, commande);
  }

  updateCommande(id: number, commande: CommandeAchat): Observable<CommandeAchat> {
    return this.http.put<CommandeAchat>(`${this.apiUrl}/${id}`, commande);
  }

  deleteCommande(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  getCommandesByFournisseur(fournisseurId: number): Observable<CommandeAchat[]> {
    return this.http.get<CommandeAchat[]>(`${this.apiUrl}/fournisseur/${fournisseurId}`);
  }

  getCommandesByStatut(statut: string): Observable<CommandeAchat[]> {
    return this.http.get<CommandeAchat[]>(`${this.apiUrl}/statut/${statut}`);
  }

  getCommandesByDateRange(dateDebut: string, dateFin: string): Observable<CommandeAchat[]> {
    return this.http.get<CommandeAchat[]>(`${this.apiUrl}/date-range?dateDebut=${dateDebut}&dateFin=${dateFin}`);
  }

  updateStatut(id: number, nouveauStatut: string): Observable<CommandeAchat> {
    return this.http.put<CommandeAchat>(`${this.apiUrl}/${id}/statut?nouveauStatut=${nouveauStatut}`, {});
  }
}

File: backend/src/main/java/com/gestionachats/service/CommandeAchatService.java
package com.gestionachats.service;

import com.gestionachats.entity.CommandeAchat;
import com.gestionachats.entity.Fournisseur;
import com.gestionachats.repository.CommandeAchatRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;

@Service
@Transactional
public class CommandeAchatService {

    @Autowired
    private CommandeAchatRepository commandeAchatRepository;

    public List<CommandeAchat> getAllCommandes() {
        return commandeAchatRepository.findAll();
    }

    public Optional<CommandeAchat> getCommandeById(Long id) {
        return commandeAchatRepository.findById(id);
    }

    public CommandeAchat saveCommande(CommandeAchat commande) {
        if (commande.getDate() == null) {
            commande.setDate(LocalDate.now());
        }
        return commandeAchatRepository.save(commande);
    }

    public CommandeAchat updateCommande(Long id, CommandeAchat commandeDetails) {
        return commandeAchatRepository.findById(id)
            .map(commande -> {
                commande.setFournisseur(commandeDetails.getFournisseur());
                commande.setDate(commandeDetails.getDate());
                commande.setStatut(commandeDetails.getStatut());
                commande.setMontant(commandeDetails.getMontant());
                return commandeAchatRepository.save(commande);
            })
            .orElse(null);
    }

    public void deleteCommande(Long id) {
        commandeAchatRepository.deleteById(id);
    }

    public List<CommandeAchat> getCommandesByFournisseurId(Long fournisseurId) {
        return commandeAchatRepository.findByFournisseurId(fournisseurId);
    }

    public List<CommandeAchat> getCommandesByFournisseur(Fournisseur fournisseur) {
        return commandeAchatRepository.findByFournisseur(fournisseur);
    }

    public List<CommandeAchat> getCommandesByStatut(String statut) {
        return commandeAchatRepository.findByStatut(statut);
    }

    public List<CommandeAchat> getCommandesByDateRange(LocalDate dateDebut, LocalDate dateFin) {
        return commandeAchatRepository.findByDateBetween(dateDebut, dateFin);
    }

    public List<CommandeAchat> getRecentCommandesByFournisseur(Long fournisseurId) {
        return commandeAchatRepository.findRecentCommandesByFournisseur(fournisseurId, LocalDate.now().minusMonths(6));
    }

    public CommandeAchat updateStatut(Long id, String nouveauStatut) {
        return commandeAchatRepository.findById(id)
            .map(commande -> {
                commande.setStatut(nouveauStatut);
                return commandeAchatRepository.save(commande);
            })
            .orElse(null);
    }
}
